Added count_members, list_members functions.

PiotrWeppo opened the pull request:
Nowe dwie funkcje na czacie Discorda:
- count_members
- list_members

Do obu można podać dowolną ilość argumentów przedzielonych operatorami and, or, not. Operatory można używać w grupach: and i not, or i not, lub dowolne same. W tym momencie nie można łączyć operatorów or i and, występuje też warunek, że najpierw muszą się znaleźć wszystkie operatory and lub wszystkie operatory or, później wszystkie operatory not.

Przykłady:
![image](https://user-images.githubusercontent.com/117805804/218268081-ecd6edc7-a1f1-4d87-aeb2-e8277f608023.png)

![image](https://user-images.githubusercontent.com/117805804/218268190-8598c87c-446b-43e5-ae4d-e5ef55d1b429.png)

![image](https://user-images.githubusercontent.com/117805804/218268004-27075b4b-8f29-4fbb-8fbf-69869c024eb9.png)

PiotrWeppo committed:
Added count_members, list_members functions.

pktiuk commented:
Czemu dodałeś nowe komendy do bota?

PiotrWeppo commented:
> Czemu dodałeś nowe komendy do bota?

Mieliśmy spotkanie, zdecydowaliśmy jak mają wyglądać dwie pierwsze komendy i egzekwuję kod.
Jeżeli funkcje będą samoistnie dobrze działać, kolejnym krokiem jest rozwój kolejnych funkcji, albo integracja z generowaną wiadomością, ale pierwszym krokiem jest zbudowanie samego podstawowego bloku funkcyjnego.

pktiuk commented:
W takim razie funkcja generująca listę osób powinna generować część wiadomości zawierającą listę osób spełniających rolę, wtedy po anpisaniu tej funkcji raz nie musiałbyś jej już dotykać, a teraz nawet jeśli funkcja działa dobrze to potem będziesz musiał ją przepisać, aby móc coś sprawdzić, poza tym ten kod nie pokazuje w jaki sposób zamierzasz parsować wiadomości.

Jak chcesz możemy usiąść i ogarnąć, aby było tak jak powinno. Zajęłoby nam to pewnie raptem kilkanaście minut.

PiotrWeppo commented:
"W takim razie funkcja generująca listę osób powinna generować część wiadomości zawierającą listę osób spełniających rolę, wtedy po anpisaniu tej funkcji raz nie musiałbyś jej już dotykać"

Funkcja generuje część wiadomości, proszę spójrz raz jeszcze na dołączone zdjęcia.

"a teraz nawet jeśli funkcja działa dobrze to potem będziesz musiał ją przepisać"

Zamienienie ją na funkcję działającą wewnątrz wiadomości to prawdopodobnie usunięcie dwóch wierszy z początku i końca funkcji, a chciałem, żebyś mógł zobaczyć rzeczywistą funkcjonalność.

"aby móc coś sprawdzić"

Stworzenie tego segmentu jako oddzielnej funkcji właśnie pozwala na sprawdzanie jak działają poszczególne elementy, zanim się stworzy pełną wiadomość.

"poza tym ten kod nie pokazuje w jaki sposób zamierzasz parsować wiadomości."

Parsowanie zostało omówione na poprzednim spotkaniu i umieszczone w README. Co prawda zaktualizowany readme nie został jeszcze wysłany, ale sam pisałeś ten fragment, więc wiesz jaki jest cel.

pktiuk commented on bot.py:
```diff
@@ -39,6 +37,145 @@ async def hello(ctx):
+        if " and " in role_names_str
+        else role_names_str.split(" or ")
+    )
+    operator = "and" if " and " in role_names_str else "or"
+
+    roles = [
+        discord.utils.get(ctx.guild.roles, name=role_name) for role_name in role_names
+    ]
+    roles = [role for role in roles if role is not None]
+    if not roles:
+        await ctx.send(
+            f"Could not find any of the roles with names `{role_names_str}`."
+        )
+        return
+
+    members = set()
+    if operator == "and":
+        role_names = " and ".join(role.name for role in roles)
+        for role in roles:
+            if not members:
+                members.update(role.members)
+            else:
+                members &= set(role.members)
+    else:
+        role_names = " or ".join(role.name for role in roles)
+        for role in roles:
+            members.update(role.members)
+
+    for not_role in not_roles:
+        members -= set(not_role.members)
```
Dobrze by było, gdybyś wydzielił kod generujący listę członków do oddzielnej funkcji.
Wtedy nie będziesz miał powtarzającego się kodu w `count_members` i `list_members`

pktiuk replied:
Poza tym funkcje będą dużo krótsze i czytelniejsze.

pktiuk replied:
Ten komentarz jest wciąż aktualny.    
Tworzenie zbioru pasujących użytkowników `members` nadal powinno być wydzielone

pktiuk replied:
Proponowałbym to podzielić na 2-3 funkcje max.

> "Tworzenie zbioru pasujących użytkowników members nadal powinno być wydzielone"
>
> Rozwiń proszę. Mówisz o podfunkcji? Dlaczego akurat specyficznie to z całej funkcji musi być wydzielone?

Mówię o stworzeniu oddzielnej podfunkcji. Akurat to musi być wydzielone, ponieważ jest to fragment, który się powtarza w dwóch miejscach. Opisałem to powyżej. Jest to blok, kodu, który pojawia się dwa razy i jest skopiowany niemal 1 do jednego.

pktiuk replied:
Stworzenie jakiejś prostej funkcji, do której podawałbyś stringa, a która zwracałaby listę ludzi pasujących do podanych kryteriów bardzo uprościłaby i skróciła twój kod

pktiuk reviewed (changes requested):
Poza tym jednym komentarzem co do kodu to nie mam większych zastrzeżeń

PiotrWeppo committed:
Extended list_members and count_members functions to allow starting messages with "not" operators, and also searching for roles using only "not" operators

PiotrWeppo committed:
Work in progress

PiotrWeppo committed:
Added function that parses list_members function inside the code.

PiotrWeppo commented:
Dodałem nowe commity. Mam problemy z łączeniem zwykłych funkcji z funkcjami asynchronicznymi, dlatego nie byłem na razie w stanie przerzucić kodu na zewnątrz. Podobnie, mam problem z przerzuceniem danych z jednej funkcji asynchronicznej do drugiej, jeżeli chodzi natomiast o samą funkcjonalność "parsowania" w tekście, to w podstawowej formie działa.

pktiuk commented on bot.py:
```diff
@@ -40,7 +37,316 @@ async def hello(ctx):
+
+
+@bot.command()
+async def searching_for_roles(ctx, output_string):
+    """Docstring placeholder"""
```
Co robi ta funkcja?   
Na pierwszy rzut oka pomyślałbym, że to ona ma filtrować listę użytkowników, ale widzę tu kilka rzeczy, które świadczą inaczej. (położenie w kodzie, jest to komenda dla bota, przyjmuje dziwne argumenty...)

PiotrWeppo replied:
jest to w prosty sposób zmieniona funkcja list_members. Zmieniona w taki sposób, żeby zamiast przyjmować argument z czatu discorda, argumentem powinna być przekonwertowana wiadomość z poprzedniej funkcji, czyli converting_string. Pierwszy argument, czyli ctx, z tego co wiem jest potrzebny, jeżeli chce się wyciągać informacje z serwera. Dodatkowo, zmieniłem zwracanie argumentów, żeby zamiast wiadomości końcowe były wysyłane do czatu, powinny być zwracane wewnątrz kodu, tam gdzie będę potem chciał.

pktiuk replied:
Zamykam, bo widzę, że już jej nie ma

PiotrWeppo committed:
Fixed issues.
Added count_members command to the parsing function.
Parsed message can be read in the embed when typed !embed in discord chat.

PiotrWeppo commented:
Tak to teraz wygląda.
![Code_ixLMrEYOFZ](https://user-images.githubusercontent.com/117805804/221379164-5998bc28-c23f-4e0a-8e26-bab4810b3d32.png)
![Discord_1OGJVApIiU](https://user-images.githubusercontent.com/117805804/221379167-fae8afa0-e0fc-4b85-8a22-74b6cafc1e4e.png)

Muszę jeszcze pomysleć jakie powinny być odpowiedzi gdy nie znajdzie jakieś roli, albo przy jakimś innym błędzie

pktiuk commented on bot.py:
```diff
@@ -39,6 +36,444 @@ async def hello(ctx):
+                f"The members without roles `{not_role_names}` are: {member_names}."
+            )
+        else:
+            await ctx.send(
+                f"The members with the roles `{role_names}`"
+                f" but not the roles `{not_role_names}` are: {member_names}."
+            )
+    else:
+        await ctx.send(
+            f"The members with the roles `{role_names}` are: {member_names}."
+        )
+"""
+
+"""
+def converting_new(input_new):
+    converted_input = f"{input_new} + new segment"
+    return converted_input
+
+
+def adding_owner(ctx, converted_input):
+    owner = ctx.guild.owner.mention
+    final_message = f"{converted_input} + {owner}"
+    return final_message
+
+
+@bot.command()
+async def show_final_message_in_embed(ctx):
+    input_new = "12345"
+    converted_input = converting_new(input_new)
+    final_message = adding_owner(ctx, converted_input)
+
+    embed_new = discord.Embed(
+        title="Info",
+        description="Info",
+        color=discord.Colour.blue(),
+    )
+    embed_new.add_field(name="Final message", value=f"{final_message}", inline=True)
+    await ctx.send(embed=embed_new)
+"""
+
```
Ten zakomentowany kod ma tu być, czy też zamierzasz go do czegoś wykorzystać później?

pktiuk commented on bot.py:
```diff
@@ -39,6 +36,444 @@ async def hello(ctx):
+                f"The members without roles `{not_role_names}` are: {member_names}."
+            )
+        else:
+            await ctx.send(
+                f"The members with the roles `{role_names}`"
+                f" but not the roles `{not_role_names}` are: {member_names}."
+            )
+    else:
+        await ctx.send(
+            f"The members with the roles `{role_names}` are: {member_names}."
+        )
+"""
+
+"""
+def converting_new(input_new):
+    converted_input = f"{input_new} + new segment"
+    return converted_input
+
+
+def adding_owner(ctx, converted_input):
+    owner = ctx.guild.owner.mention
+    final_message = f"{converted_input} + {owner}"
+    return final_message
+
+
+@bot.command()
+async def show_final_message_in_embed(ctx):
+    input_new = "12345"
+    converted_input = converting_new(input_new)
+    final_message = adding_owner(ctx, converted_input)
+
+    embed_new = discord.Embed(
+        title="Info",
+        description="Info",
+        color=discord.Colour.blue(),
+    )
+    embed_new.add_field(name="Final message", value=f"{final_message}", inline=True)
+    await ctx.send(embed=embed_new)
+"""
+
```
Ten zakomentowany kod ma tu być? Czy też zapomniałeś go usunąć?

PiotrWeppo replied:
To, a przynajmniej funkcje list_members i count_members, to początkowe funkcje, które działają w bocie i zastanawiałem się czy je zostawić zakomentowane, jakbym chciał szybko testować jakieś zmiany w funkcjach, ale widzę, że nie podobają się linterom ze względu na długość, więc w kolejnej wersji je usunę.

pktiuk replied:
Masz na myśli kolejną wersję bota, czy kolejną wersję tego brancha?

pktiuk commented on bot.py:
```diff
@@ -39,6 +36,444 @@ async def hello(ctx):
+
+
+def counting_members(ctx, second_string):
+    """Docstring placeholder"""
```
DObrze by było usunąć, z kodu wszelkie placeholdery (albo w wypadku nieoczywistych funkcji napisać jakieś komentarze)

PiotrWeppo replied:
Nie było to moim priorytetem w tym czasie, więc mogłem wybrać między tym co jest, a zostawienie pustych, ale miałbym wtedy masę ostrzeżeń od pylint'a. Jak posprzątam funkcje, to zajmę się tym.

pktiuk reviewed (changes requested):
Kod jest już lepszy, mam jeszcze kilka uwag w poszczególnych wątkach.

pktiuk commented:
W jednym z wątków nadal nie dałeś odpowiedzi

PiotrWeppo commented:
> W jednym z wątków nadal nie dałeś odpowiedzi

nie widzę, żadnego innego wątku

pktiuk commented:
![image](https://user-images.githubusercontent.com/45544416/221642231-595cf53f-db0b-4638-8f87-ab8600b7cf47.png)

PiotrWeppo commented:
Tak zrobię. Mam pytanie: zacząłem ładować te dwie wielkie funkcje listing_members i counting_members w mniejsze wspólne funkcje na zewnątrz i podzieliłem sobie z grubsza komentarzami sektory, które wchodziłyby w skład oddzielnych funkcji, ale zastanawiam się, czy jak w tych funkcjach będzie nagle zanestowane 7 wywoływań innych funkcji to czy to na pewno będzie czytelne. Zrobić tak, czy zredukować może do 2-3 podfunkcji max?
![1](https://user-images.githubusercontent.com/117805804/221643754-e35e31e8-56c0-4ef2-8278-7bef2d531f6d.png)
![2](https://user-images.githubusercontent.com/117805804/221643790-c4f0360a-6f59-409b-ab4e-7f37d25da614.png)
![3](https://user-images.githubusercontent.com/117805804/221643816-a6d34557-7378-40f6-896b-d773958f21db.png)

od jednego zielonego komenta do drugiego to z grubsza teoretyczna podfunkcja

PiotrWeppo commented:
"Tworzenie zbioru pasujących użytkowników members nadal powinno być wydzielone"

Rozwiń proszę. Mówisz o podfunkcji? Dlaczego akurat specyficznie to z całej funkcji musi być wydzielone?

pktiuk commented:
Odpisałem ci w wątku (i ty też odpisuj mi w poszczególnych wątkach, aby unikać spamu)

pktiuk commented:
Zamykam, zamiast niego zmerguję https://github.com/KNR-PW/discord-bot/pull/13
